### ISSUE TITLE
Code comments

### ISSUE THREAD

**Issue by arogozhnikov** [RosettaCommons/RFdiffusion#29]
Hi team, congratulation on the great paper and milestone results.

I've collected multiple remarks on the code while I was studying it, hopefully this will help in improving the codebase.

**terminology** 

- templates dimension: sometimes called T, sometimes N, sometimes s.
- t1d and t2d are very non-descriptive names. 

**dependencies**
- opt_einsum is used, but should be replaced with torch.einsum: 1. pytorch relies on opt_einsum anyway 2. opt_einsum handles contraction order for 3 or more tensors, which is never the case in rosetta code  3. it precludes scripting/compiling code

**ignored parameters/code**
- igso3.py: calculate_igso3 ignores L argument
- init_lecun_normal and init_lecun_normal_param ignore passed initialization scale 
- Attention and MaskedTokenNetwork ignore p_drop, and dropout is never applied (while signature suggests the opposite)
- class Denoise: multiple parameters are not used
- looks like `* self.guide_scale` is missing here https://github.com/RosettaCommons/RFdiffusion/blob/92b83decf3b1fd18794fc5f8a7a94d4124035fc8/rfdiffusion/potentials/manager.py#L202
- diff_util.py: unused file
- olig_intra_contacts: misses self.d0 and self.r0

**distributing weights**

- I'd recommend replacing unsafe pickles with something simple and safe (npz/safetensors)

**computations**
- manual normalizations (lots of them) can be replaced with F.normalize, cosines (in many places) can be computed with F.cosine_similarity
- `Sergey's one hot trick` (used in multiple places) - strangely, created embedding layer is never used. Better just create a linear module, or only a trainable parameter
- looks like you try to implement torch.nanmean here https://github.com/RosettaCommons/RFdiffusion/blob/92b83decf3b1fd18794fc5f8a7a94d4124035fc8/rfdiffusion/kinematics.py#L293-L295 

**checkpointing**

- Checkpointing on iterblock: wouldn’t it be more efficient to checkpoint whole block instead of checkpointing every step? This will simplify logic as well
- create_custom_forward: unclear why you need it. This function was used only once with an actual kwarg (topk), and it can be replaced with lambda anyway: `lambda *args: module(*args, topk=...)`

**duplicated code**

- computation of Cbeta: three torch implementations in repo, **and** there is discrepancy in used weights
- make_contact_matrix: implemented twice
- dihedral computation: implemented twice in numpy, once in pytorch, implementations scattered

Minor:
- PositionalEncoding2D: I think binning is asymmetric (enumerates shifts from -32 to 31 + outside regions instead of -32 to 32)
- icecream is imported but never used (and required in dependencies)

**Comment by sokrypton** [comment]
Thanks @arogozhnikov! this is very helpful

**Comment by joewatchwell** [comment]
Thanks for this - we will go through the code and make the changes :)

**Comment by joewatchwell** [comment]
We've now made most of these changes - thanks again!
